### ISSUE TITLE
CustomMapping causes problem when used inside an Autodesk plugin

### ISSUE THREAD

**Issue by ntpthinh** [dotnet/machinelearning#5405]
### System information

- **OS version/distro**: Windows
- **.NET Version (eg., dotnet --info)**:  .NET framework 4.7

### Issue

- **What did you do?**
I use [Normalize CustomMapping](https://github.com/dotnet/machinelearning-samples/blob/master/samples/modelbuilder/ImageClassification_Azure_LandUse/LandUseML.Model/NormalizeMapping.cs) for ONNX model 
Without the mapping, everything works fine and it stills worked in development environment. The model, the way I load the model are the same.

- **What happened?**
When I load my application as a plugin for another application (with all of needed libraries), they always ask for missing library although they have been added (exact version). New libraries are asked each time but all of them are releated to ML.NET (Microsoft.ML.Transforms, System.Memory for eg.). I wonder if CustomMapping causes incompatibility in library version. 
- **What did you expect?**
I can load library without errors.
### Source code / logs
The code is basically the same as tutorial but I will show here for validation
```
MLContext mlContext = new MLContext();
var data = mlContext.Data.LoadFromEnumerable(new List<ImageData>());
var pipeline = mlContext.Transforms.LoadImages(outputColumnName: "image", imageFolder: "", inputColumnName: nameof(ImageData.ImagePath))
                    .Append(mlContext.Transforms.ResizeImages(outputColumnName: ModelSettings.ModelInput, imageWidth: ImageResNetSettings.imageWidth, imageHeight: ImageResNetSettings.imageHeight, inputColumnName: "image"))
                    .Append(mlContext.Transforms.ExtractPixels(outputColumnName: ModelSettings.ModelInput))
                    .Append(mlContext.Transforms.CustomMapping(new NormalizeMapping().GetMapping(), contractName: nameof(NormalizeMapping)))
                    .Append(mlContext.Transforms.ApplyOnnxModel(modelFile: modelLocation, outputColumnNames: new[] { ModelSettings.ModelOutput }, inputColumnNames: new[] { ModelSettings.ModelInput }));
var model = pipeline.Fit(data);
mlContext.ComponentCatalog.RegisterAssembly(typeof(NormalizeMapping).Assembly);
mlContext.Model.Save(model, data.Schema, _outputModelFilePath);
```

```Uri uri = new Uri(modelLocation, UriKind.Relative);
StreamResourceInfo info = Application.GetResourceStream(uri);                  
trainedModel = mlContext.Model.Load(info.Stream, out DataViewSchema modelSchema);
```

**Comment by antoniovs1029** [comment]
Can you please share the exception message you receive and the full stack trace?
Also, if possible, can you share a .zip with your full code and dataset?
Thanks

**Comment by pthinh14** [comment]
The error is ```System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.ML.Transforms, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
File name: 'Microsoft.ML.Transforms, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'```.
When I try to add the dll explicitly, it keeps saying other dll are missing so probably the real exception is being eaten (?)


Please refer to https://github.com/pthinh14/Plugin-Test for source code. I created a sufficient version to check that error. 
I'm pretty sure the model is fine enough. I work perfectly when there is no CustomMapping (wrong result though) or when I try add Microsoft.ML.Transforms.dll explicitly (although it already included in the application)

**Comment by antoniovs1029** [comment]
Unfortunately, I wasn't able to run your solution since it requires some Autodesk nugets which my Visual Studio don't seem to get automatically from nuget.org, so my guess is that those nugets need to be taken from some other source. Also, your repo doesn't include the .onnx or model.zip file related to your project. So I wasn't able to run it. Furthermore, even if it ran, I guess I wouldn't be able to test the plugin on Autodesk, , since I don't have access to Autodesk, which is where you seem to be getting those exceptions.

Overall, from looking at your code I don't see anything that could cause a problem. But, to be fair, I don't know how Autodesk plugins are built either, so maybe the problem lies in how to setup your project, and not on ML.NET itself.

Still, I would like to make 2 suggestions, so that you can try them out to see if they work out:

1. **If you train a model everytime Button_Click runs, then you can avoid using CustomMappingFactoryAttribute**. It is unclear to me what should happen when `Button_Click` executes. On your current code, it says it should train a model, but there are some commented lines of code where it would load a model instead. If it's ok to train a model everytime the `Button_Click` method runs, then you don't need to use `mlContext.ComponentCatalog.RegisterAssembly()`, and so you wouldn't need to have a class with the `CustomMappingFactoryAttribute()`, as you could define the custom mapping as a simple delegate. Please refer to [this sample](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs) to see how to create a CustomMappingTransformer without a CustomMappingFactoryAttribute() and without RegisterAssembly. For reference, you can compare that sample with [this other sample](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs) that applies the same custom transformation but with CustomMappingFactoryAttribute() and RegisterAssembly(). Please notice that the attribute and RegisterAssembly() are only needed when you have to save/load a model.zip containing the custom transformer.

or

2. **Try using ML.NET 1.4.0 .** On ML.NET 1.5.0 there was a change on how assemblies are loaded with CustomMappingTransformer (see PR #4989 ). I don't think that change could cause the exception you're receiving, but it's worth trying out 1.4.0 to see if it helps in anything. Notice that if you try to load a model.zip created with 1.5.0 on 1.4.0 it might fail, so it would be necessary to train and save again the model, but only using 1.4.0.

From the behavior you describe, I hope that my suggestion # 1 might help. My guess is that there's a problem in how you're building and/or packaging the plug-in, that throws and error when trying to use the `Microsoft.ML.Transforms.CustomMappingFactoryAttribute` on your code because, for whatever reason, it can't find the Microsoft.ML.Transforms .dll. So maybe getting rid of the attribute would help in not facing this problem.

If the problem persists, please, share the following info:
1. Complete stack trace of your exception. You've shared the exception message, but the full stack trace might be useful to further investigate this.
2. If you're loading a model.zip file, please share the model.zip.
3. Can you please give more details about this:
> I'm pretty sure the model is fine enough. I work perfectly when there is no CustomMapping (wrong result though) or when I try add Microsoft.ML.Transforms.dll explicitly (although it already included in the application)

What do you mean when you say you "try to add the dll explicitly" and when you say that "it already included in the application"? If it was _already_ included in your plugin, then how did you add the dll explicitly?

Thanks

**Comment by ntpthinh** [comment]
Thank you for looking into my problem.

Sorry I didn't make it clear enough. Actually, I need to save the model and load it in a different place because mlContext.Transforms.ApplyOnnxModel only allow model file path and the model is attached in my dll. 
The code is to show what I have done with saving and loading process. In my delivery, I use the loading part and CustomMappingFactoryAttribute to load the model. In here, I just want to show that as long as CustomMappingFactoryAttribute  is presented, it fails to load the library.

> `What do you mean when you say you "try to add the dll explicitly" and when you say that "it already included in the application"? If it was already included in your plugin, then how did you add the dll explicitly?`
When I check my delivery dll, all the necessary libraries are presented but Autodesk keeps asking that Microsoft.ML.Transforms.dll is missing. Then when I add Microsoft.ML.Transforms.dll explicitly to Autodesk , it can found Microsoft.ML.Transforms.dll but asking for other missing dlls.
Please note that this will not happen without CustomMappingFactoryAttribute.

I already try to downgrade ML.NET to 1.4.0 but it doesn't seem to work.
Here is full stack trace that I got:
![image](https://user-images.githubusercontent.com/20205286/94523964-2a02b780-025c-11eb-9224-57c5066d0be5.png)

**Comment by ntpthinh** [comment]
Is there anyway to load onnx model as a stream not by model location? If yes my problem can be resolved. Thank you!

**Comment by antoniovs1029** [comment]
Thanks for your answers and for trying out 1.4.0. Since using 1.4.0 didn't work, then I don't think that's the problem, and it's ok that you continue to try to fix this problem using 1.5.2.

> Actually, I need to save the model and load it in a different place because mlContext.Transforms.ApplyOnnxModel only allow model file path and the model is attached in my dll. The code is to show what I have done with saving and loading process. In my delivery, I use the loading part and CustomMappingFactoryAttribute to load the model. In here, I just want to show that as long as CustomMappingFactoryAttribute is presented, it fails to load the library.

I don't think the problem is with loading the .onnx model, because `CustomMappingFactoryAttribute` is used to load the ML.NET model.zip model to load the `CustomMappingTransformer`... the `CustomMappingFactoryAttribute` is NOT needed to load the .onnx model. Furthermore, you said you were able to run `ApplyOnnxModel` when you didn't used `CustomMappingTransformer`, right? So this also means that the problem isn't with loading the .onnx model.

So the real question isn't if you need to load your .onnx model, but if you need to save and load your ML.NET model.zip model. If you don't need to save and load the ML.NET model then, as I've suggested, you can get rid of the `CustomMappingFactoryAttribute` and hopefully this problem would go away.

> Here is full stack trace that I got:

From your stack trace, I don't know what the `AssemblyLoader` class is, but I'd think it's a component on AutoDesk side to load your plugin. Also, from the behavior you're describing, I still think the problem isn't in ML.NET (i.e. this isn't a bug or an issue on ML.NET), but rather on how the plugin is being created. Because of this, I'd think you'd be able to find more information on how to solve this problem by reaching out to the Autodesk community, as they might be more informed to help you out with this problem.

> Is there anyway to load onnx model as a stream not by model location? If yes my problem can be resolved

Unfortunately, ML.NET doesn't support loading onnx models from streams, only from a path.

As I've mentioned, since you didn't have problems loading the .onnx model before using `CustomMappingTransformer`, then I don't think this would solve your problem either. Even if you were able to load the onnx model from a stream, you'd still need the `CustomMappingTransformer` to do normalization, so the problem wouldn't be fixed.

Still, if for whatever reason you want to load an onnx model from stream, then you can try to do it with [Onnxruntime](https://github.com/Microsoft/onnxruntime/blob/master/docs/CSharp_API.md). I wasn't able to find a tutorial that explains how to do it, but they do provide this method to create an Onnx session using a byte stream with the model:

https://github.com/microsoft/onnxruntime/blob/11c194ce290eb6309c44c1ecf1991a84f26dbf51/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs#L53

If you decide to use Onnxruntime instead of ML.NET, please just realize that you'll need to write the code to do the preprocessing (loading images, resizing images, extracting pixels, and normalizing), as Onnxruntime doesn't work with ML.NET's IDataViews, but with standard C# arrays. I don't know if there are tools on Onnxruntime to do this preprocessing.

> Please note that this will not happen without CustomMappingFactoryAttribute.

Have you actually tried my suggestion [# 1 above](https://github.com/dotnet/machinelearning/issues/5405#issuecomment-699873935)?

Typically training ML.NET models takes a lot more time than loading pretrained model.zip models. But on your scenario, I don't think that's the case, because your pipeline doesn't need training, so I think you can get away without loading a model.zip model, and simply defining the pipeline every time you run `Button_Click`. So I think doing this would work (I haven't tried it, since I can't run your solution, but the code should look something like this):

```C#
        private void Button_Click(object sender, RoutedEventArgs e)
        {
            Action<NormalizeInput, NormalizeOutput> mapping =
                (input, output) =>
                {
                    var values = input.Reshape.GetValues().ToArray();

                    var image_mean = new float[] { 0.5f, 0.5f, 0.5f };
                    var image_std = new float[] { 0.5f, 0.5f, 0.5f };

                    for (int x = 0; x < values.Count(); x++)
                    {
                        var y = x % 3;
                        // Normalize by 255 first
                        values[x] /= 255;
                        values[x] = (values[x] - image_mean[y]) / image_std[y];
                    };

                    output.Reshape = new VBuffer<float>(values.Count(), values);
                };


            try
            {
                string modelLocation = @"D:\model.onnx";

                MLContext mlContext = new MLContext();
                var data = mlContext.Data.LoadFromEnumerable(new List<ImageData>());
                var pipeline = mlContext.Transforms.LoadImages(outputColumnName: "image", imageFolder: "", inputColumnName: nameof(ImageData.ImagePath))
                        .Append(mlContext.Transforms.ResizeImages(outputColumnName: ModelSettings.ModelInput, imageWidth: ImageResNetSettings.imageWidth, imageHeight: ImageResNetSettings.imageHeight, inputColumnName: "image"))
                        .Append(mlContext.Transforms.ExtractPixels(outputColumnName: ModelSettings.ModelInput))
                        .Append(mlContext.Transforms.CustomMapping(mapping, contractName: null))
                        .Append(mlContext.Transforms.ApplyOnnxModel(modelFile: modelLocation, outputColumnNames: new[] { ModelSettings.ModelOutput }, inputColumnNames: new[] { ModelSettings.ModelInput }));
                var model = pipeline.Fit(data);

                IEnumerable<ImageData> image = ImageData.ReadImageFromPaths(new string[] { @"E:\test.jpg" });
                IDataView imageDataView = mlContext.Data.LoadFromEnumerable(image);
                IDataView scoredData = model.Transform(imageDataView);
                List<float[]> probabilities = scoredData.GetColumn<float[]>(ModelSettings.ModelOutput).ToList();
            }
            catch(Exception ex) { MessageBox.Show(ex.Message); }
        }

// Remove the NormalizeMapping class.
```

By defining the mapping as an `Action<,>` delegate, you don't need to define the `NormalizeMapping` class, and so you won't need to use `CustomMappingFactoryAttribute`, you also don't need to provide a `contractName`, nor use `RegisterAssembly()`.

As I mentioned before, this approach won't work if you need to save and load an ML.NET model.zip, but in your case it doesn't seems necessary to do that, and so I hope the code I've shared will work.

Please, let us know if you've tried this code, and if it worked for you. Thanks.

**Comment by antoniovs1029** [comment]
After the assessment I provided above it seems that the problem is in how the plugin is being created, so I will close this issue now.
